File: go/tencent/backtrack/generate_parenthesis.go
package main

import (
	"container/list"
	"fmt"
)

func main() {
	fmt.Println(generateParenthesis(3))
}

func generateParenthesis(n int) []string {
	res := list.New()
	recursion(res, "", n, n)
	returnVal := make([]string, res.Len())
	for e, i := res.Front(), 0; e != nil; e, i = e.Next(), i+1 {
		returnVal[i] = (e.Value).(string)
	}
	return returnVal
}

func recursion(res *list.List, temp string, left, right int) {
	if left == 0 && right == 0 {
		res.PushBack(temp)
		return
	}
	if left > 0 {
		recursion(res, temp+"(", left-1, right)
	}
	if right > 0 && right > left {
		recursion(res, temp+")", left, right-1)
	}
}

File: java/leetcode-explore/highFrequency/backTrack/GenerateParenthesis.java
package highFrequency.backTrack;

import java.util.ArrayList;
import java.util.List;

public class GenerateParenthesis {

    public List<String> generateParenthesis(int n) {
        List<String> res = new ArrayList<>();
        recursion(n, n, "", res);
        return res;
    }

    public void recursion(int left, int right, String tmp, List<String> res) {
        if (left == 0 && right == 0) {
            res.add(tmp);
            return;
        }
        if (left > 0) {
            recursion(left - 1, right, tmp + "(", res);
        }

        if (right > 0 && right > left) {
            recursion(left, right - 1, tmp + ")", res);
        }
    }
}
